cmd/metal-api/internal/scaler/poolscaler.go

```diff
@@ -0,0 +1,178 @@
+package scaler
+
+import (
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	"go.uber.org/zap"
+)
+
+type (
+	PoolScaler struct {
+		log     *zap.SugaredLogger
+		manager MachineManager
+	}
+	PoolScalerConfig struct {
+		Log     *zap.SugaredLogger
+		Manager MachineManager
+	}
+	MachineManager interface {
+		AllMachines() (metal.Machines, error)
+		WaitingMachines() (metal.Machines, error)
+		ShutdownMachines() (metal.Machines, error)
+
+		Shutdown(m *metal.Machine) error
+		PowerOn(m *metal.Machine) error
+	}
+)
+
+func NewPoolScaler(log *zap.SugaredLogger, manager MachineManager) *PoolScaler {
+	return &PoolScaler{
+		log:     log,
+		manager: manager,
+	}
+}
+
+// AdjustNumberOfWaitingMachines compares the number of waiting machines to the required pool size of the partition and shuts down or powers on machines accordingly
+func (p *PoolScaler) AdjustNumberOfWaitingMachines(partition *metal.Partition) error {
+	waitingMachines, err := p.manager.WaitingMachines()
+	if err != nil {
+		return err
+	}
+
+	poolSizeExcess := 0
```

Comment:
This seems to be an ineffective assignment, it's always overwritten by the next line.